src/utils/SpokeUtils.ts

```diff
@@ -0,0 +1,142 @@
+import { SpokePool } from "../typechain";
+
+/**
+ * Find the block range that contains the deposit ID. This is a binary search that searches for the block range
+ * that contains the deposit ID.
+ * @param targetDepositId The target deposit ID to search for.
+ * @param initLow The initial lower bound of the block range to search.
+ * @param initHigh The initial upper bound of the block range to search.
+ * @param maxSearches The maximum number of searches to perform. This is used to prevent infinite loops.
+ * @returns The block range that contains the deposit ID.
+ * @note  // We want to find the block range that satisfies these conditions:
+ *        // - the low block has deposit count <= targetDepositId
+ *        // - the high block has a deposit count > targetDepositId.
+ *        // This way the caller can search for a FundsDeposited event between [low, high] that will always
+ *        // contain the event emitted when deposit ID was incremented to targetDepositId + 1. This is the same transaction
+ *        // where the deposit with deposit ID = targetDepositId was created.
+ */
+export async function getBlockRangeForDepositId(
```

Comment:
One consideration - it might be nice to wrap this method in another that accepts an array of deposits for the origin SpokePool. The reason for this is that we're likely to call `getBlockRangeForDepositId()` in parallel, and so there will be many concurrent calls to `getBlockNumber()` (etc etc). When these are executed in parallel for the first time, they're unlikely to be served out of the local redis cache, so we'll probably hit the RPC provider with a bunch of identical requests.

It might therefore be nice if we resolve that once centrally, and supply the values in as optional arguments. If the caller has access to a `SpokePoolClient` instance then they could then supply `SpokePoolClient.latestBlockNumber`, and that would save a lookup.

Reply:
I agree. In this scenario, do you envision the multi-id function to be the main & the current existing process to be the one-off or vice versa?

Reply:
I was thinking that we'd collate a list of deposit Ids that we're interested in, and then query them all at once. It might require further refactoring upstream in the calling code though, so perhaps it's not practical given the current code structure.